File: src/main/java/com/atsuishio/superbwarfare/capability/player/PlayerVariable.java
package com.atsuishio.superbwarfare.capability.player;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.data.gun.Ammo;
import com.atsuishio.superbwarfare.init.ModAttachments;
import com.atsuishio.superbwarfare.network.message.receive.PlayerVariablesSyncMessage;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.common.util.INBTSerializable;
import net.neoforged.neoforge.event.entity.player.PlayerEvent;
import net.neoforged.neoforge.network.PacketDistributor;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;


@EventBusSubscriber(modid = Mod.MODID)
public class PlayerVariable implements INBTSerializable<CompoundTag> {
    private PlayerVariable old = null;

    public Map<Ammo, Integer> ammo = new EnumMap<>(Ammo.class);
    public boolean activeThermalImaging = false;

    public void sync(Entity entity) {
        if (!entity.hasData(ModAttachments.PLAYER_VARIABLE)) return;

        var newVariable = entity.getData(ModAttachments.PLAYER_VARIABLE);
        if (old != null && old.equals(newVariable)) return;

        if (entity instanceof ServerPlayer serverPlayer) {
            PacketDistributor.sendToPlayer(serverPlayer, new PlayerVariablesSyncMessage(entity.getId(), compareAndUpdate()));
        }
    }

    public static void modify(Player player, Consumer<PlayerVariable> consumer) {
        var cap = player.getData(ModAttachments.PLAYER_VARIABLE).watch();
        consumer.accept(cap);
        cap.sync(player);
    }

    public static PlayerVariable getOrDefault(Entity entity) {
        return entity.getData(ModAttachments.PLAYER_VARIABLE);
    }

    @SubscribeEvent
    public static void onPlayerLoggedIn(PlayerEvent.PlayerLoggedInEvent event) {
        if (!(event.getEntity() instanceof ServerPlayer player)) return;

        PacketDistributor.sendToPlayer(player, new PlayerVariablesSyncMessage(player.getId(), getOrDefault(player).compareAndUpdate()));
    }

    @SubscribeEvent
    public static void onPlayerRespawn(PlayerEvent.PlayerRespawnEvent event) {
        if (!(event.getEntity() instanceof ServerPlayer player)) return;

        PacketDistributor.sendToPlayer(player, new PlayerVariablesSyncMessage(player.getId(), getOrDefault(player).compareAndUpdate()));
    }

    @SubscribeEvent
    public static void onPlayerChangeDimension(PlayerEvent.PlayerChangedDimensionEvent event) {
        if (!(event.getEntity() instanceof ServerPlayer player)) return;

        PacketDistributor.sendToPlayer(player, new PlayerVariablesSyncMessage(player.getId(), getOrDefault(player).forceUpdate()));
    }

    public PlayerVariable watch() {
        this.old = this.copy();
        return this;
    }

    public Map<Byte, Integer> forceUpdate() {
        var map = new HashMap<Byte, Integer>();

        for (var type : Ammo.values()) {
            map.put((byte) type.ordinal(), type.get(this));
        }

        map.put((byte) -1, this.activeThermalImaging ? 1 : 0);


        return map;
    }

    public Map<Byte, Integer> compareAndUpdate() {
        var map = new HashMap<Byte, Integer>();
        var old = this.old == null ? new PlayerVariable() : this.old;

        for (var type : Ammo.values()) {
            var oldCount = old.ammo.getOrDefault(type, 0);
            var newCount = type.get(this);

            if (oldCount != newCount) {
                map.put((byte) type.ordinal(), newCount);
            }
        }

        if (old.activeThermalImaging != this.activeThermalImaging) {
            map.put((byte) -1, this.activeThermalImaging ? 1 : 0);
        }

        return map;
    }


    public CompoundTag writeToNBT() {
        CompoundTag nbt = new CompoundTag();

        for (var type : Ammo.values()) {
            type.set(nbt, type.get(this));
        }

        nbt.putBoolean("ActiveThermalImaging", activeThermalImaging);

        return nbt;
    }

    public void readFromNBT(CompoundTag tag) {
        for (var type : Ammo.values()) {
            type.set(this, type.get(tag));
        }

        activeThermalImaging = tag.getBoolean("ActiveThermalImaging");
    }

    public PlayerVariable copy() {
        var clone = new PlayerVariable();

        for (var type : Ammo.values()) {
            type.set(clone, type.get(this));
        }

        clone.activeThermalImaging = this.activeThermalImaging;

        return clone;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof PlayerVariable other)) return false;

        for (var type : Ammo.values()) {
            if (type.get(this) != type.get(other)) return false;
        }

        return activeThermalImaging == other.activeThermalImaging;
    }

    @SubscribeEvent
    public static void clonePlayer(PlayerEvent.Clone event) {
        event.getOriginal().revive();
        var original = event.getOriginal().getData(ModAttachments.PLAYER_VARIABLE);
        if (event.getEntity().level().isClientSide()) return;

        event.getEntity().setData(ModAttachments.PLAYER_VARIABLE, original.copy());
    }

    @Override
    public CompoundTag serializeNBT(HolderLookup.@NotNull Provider provider) {
        return writeToNBT();
    }

    @Override
    @ParametersAreNonnullByDefault
    public void deserializeNBT(HolderLookup.Provider provider, CompoundTag nbt) {
        readFromNBT(nbt);
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/capability/player/PlayerVariable.kt
package com.atsuishio.superbwarfare.capability.player

import com.atsuishio.superbwarfare.Mod.Companion.loc
import com.atsuishio.superbwarfare.capability.ModCapabilities
import com.atsuishio.superbwarfare.capability.player.PlayerVariable.Companion.modify
import com.atsuishio.superbwarfare.capability.sync.CapabilitySync
import com.atsuishio.superbwarfare.capability.sync.SyncTarget
import com.atsuishio.superbwarfare.capability.sync.SyncedCapability
import com.atsuishio.superbwarfare.data.gun.Ammo
import com.atsuishio.superbwarfare.serialization.ByteBufDecoder
import com.atsuishio.superbwarfare.serialization.ByteBufEncoder
import net.minecraft.nbt.CompoundTag
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraftforge.common.capabilities.AutoRegisterCapability
import net.minecraftforge.common.util.INBTSerializable
import java.util.*
import java.util.function.Consumer

@AutoRegisterCapability
class PlayerVariable : INBTSerializable<CompoundTag>, SyncedCapability {
    private var old: PlayerVariable? = null

    @JvmField
    var ammo: MutableMap<Ammo, Int> = EnumMap(Ammo::class.java)
    var activeThermalImaging: Boolean = false

    /** 玩家变量只对本人有意义 */
    override val syncTarget: SyncTarget
        get() = SyncTarget.SELF

    /**
     * 全量同步：先写热成像状态，再按 [Ammo] 的枚举顺序写各弹种的存量。
     * 前后顺序必须与 [readSync] 一致。
     */
    override fun writeSync(encoder: ByteBufEncoder, full: Boolean) {
        encoder.encodeBoolean(activeThermalImaging)
        encoder.encodeInt(Ammo.entries.size)

        for (type in Ammo.entries) {
            encoder.encodeInt(type.get(this))
        }
    }

    override fun readSync(decoder: ByteBufDecoder, full: Boolean) {
        activeThermalImaging = decoder.decodeBoolean()

        val size = decoder.decodeInt()
        for (index in 0 until size) {
            val count = decoder.decodeInt()

            // 弹种数量以本地枚举为准，多余的数据直接丢弃，避免越界
            if (index >= Ammo.entries.size) continue

            // 直接写入 map：这里应用的是服务端的权威数据，不能再拿客户端本地的
            // SERVER 配置上限去校验（Ammo#set 会因上限不符而静默丢弃）
            ammo[Ammo.entries[index]] = count
        }
    }

    /** 记录当前状态，配合 [equals] 判断 [modify] 是否真的产生了变化 */
    fun watch(): PlayerVariable {
        this.old = this.copy()
        return this
    }

    /**
     * 是否有未同步的变化。仅在 [watch] 之后调用有意义。
     */
    fun changed(): Boolean = old != null && old != this

    fun writeToNBT(): CompoundTag {
        val nbt = CompoundTag()

        for (type in Ammo.entries) {
            type.set(nbt, type.get(this))
        }

        nbt.putBoolean("ActiveThermalImaging", activeThermalImaging)

        return nbt
    }

    fun readFromNBT(tag: CompoundTag) {
        for (type in Ammo.entries) {
            type.set(this, type.get(tag))
        }

        activeThermalImaging = tag.getBoolean("ActiveThermalImaging")
    }

    fun copy(): PlayerVariable {
        val clone = PlayerVariable()

        for (type in Ammo.entries) {
            type.set(clone, type.get(this))
        }

        clone.activeThermalImaging = this.activeThermalImaging

        return clone
    }

    override fun equals(other: Any?): Boolean {
        if (other !is PlayerVariable) return false

        for (type in Ammo.entries) {
            if (type.get(this) != type.get(other)) return false
        }

        return activeThermalImaging == other.activeThermalImaging
    }

    override fun serializeNBT(): CompoundTag {
        return writeToNBT()
    }

    override fun deserializeNBT(nbt: CompoundTag) {
        readFromNBT(nbt)
    }

    companion object {
        var ID: ResourceLocation = loc("player_variables")

        @JvmStatic
        fun getOrDefault(entity: Entity): PlayerVariable {
            return entity.getCapability(ModCapabilities.PLAYER_VARIABLE).orElse(PlayerVariable())
        }

        /**
         * 标记玩家变量已变化，由 [CapabilitySync] 在本 tick 结束时自动同步给该玩家。
         */
        @JvmStatic
        fun markDirty(entity: Entity) {
            CapabilitySync.markDirty(entity, ID)
        }

        /**
         * 编辑并自动同步玩家变量
         *
         * 数据没有实际变化时不会发包。
         */
        @JvmStatic
        fun modify(entity: Entity, consumer: Consumer<PlayerVariable>) {
            if (entity.level().isClientSide) return
            val cap = entity.getCapability(ModCapabilities.PLAYER_VARIABLE).orElse(PlayerVariable())

            cap.watch()
            consumer.accept(cap)

            if (!cap.changed()) return

            markDirty(entity)
        }
    }

    override fun hashCode(): Int {
        var result = activeThermalImaging.hashCode()
        result = 31 * result + (old?.hashCode() ?: 0)
        result = 31 * result + ammo.hashCode()
        return result
    }
}
